File: C++/BinarySearch.cpp
/*
Program to search an element in an array using Binary Search
Author: Prajjwal Pathak
Date Modified: 3rd October 2021
*/

#include <iostream>

using namespace std;

//Array input function
void Input(int *A, int N)
{
    int i;
    for (i = 0; i < N; i++)
    {
        cin >> A[i];
    }
}

//Binary Search Function
/*
A: Array
N: Array Size
E: Element
F: First
L: Last
M: Middle
*/
int Search(int *A, int N, int E)
{
    int i, M, F, L;
    F = 0;
    L = N - 1;
    while (F < L)
    {
        M = (F + L) / 2;
        if (A[M] == E)
        {
            return M;
        }
        else if (E < A[M])
        {
            L = M - 1;
        }
        else
        {
            F = M + 1;
        }
    }
    return -1;
}

int main()
{
    int n, e, s;
    cout << "Enter the size: ";
    cin >> n;
    int a[n];
    cout << "Enter a sorted Array: ";
    Input(a, n);
    cout << "Enter the element you want to search: ";
    cin >> e;
    s = Search(a, n, e); //Function call
    if (s == -1)
    {
        cout << "Element Not Found";
    }
    else
    {
        cout << "Element is at position " << s + 1;
    }

    return 0;
}

File: JAVA/BinarySearch.java
/*
  Code: Binary Search (JAVA)
  Name: Shreeyanshi Gautam
  Date of Modification: 31/10/2021
 */
import java.util.Scanner;  
class BinarySearch 
{
    int binarySearch(int arr[], int l, int r, int x)
    {
        if (r >= l) 
        {
            int mid = l + (r - l) / 2;
            if (arr[mid] == x)
                return mid;
            if (arr[mid] > x)
                return binarySearch(arr, l, mid - 1, x);
            return binarySearch(arr, mid + 1, r, x);
        }
        return -1;
    }
    public static void main(String args[])
    {
        BinarySearch obj = new BinarySearch();
        Scanner sc=new Scanner(System.in);  
        System.out.print("Enter the number of elements: ");  
        int n=sc.nextInt();  
        int[] arr = new int[10];  
        System.out.println("Enter the elements of the array in a sorted manner: ");  
        for(int i=0; i<n; i++)  
        {   
          arr[i]=sc.nextInt();  
        }  
        System.out.println("Enter the element to be searched: ");
        int x = sc.nextInt();
        int result = obj.binarySearch(arr, 0, n - 1, x);
        if (result == -1)
            System.out.println("Element not present");
        else
            System.out.println("Element found at position: "+ (result+1));
    }
}
